### PR TITLE
feat(frontend): support TopN with ties

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

See Release note for the change.

Details:
- frontend part is straightforward
- executor: do not include `group_key` for Group TopN in cache. This fixes `order_by_len` logic for group top n
- fixed an edge case in `TopNCache::<true>::insert`

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* SQL commands, functions, and operators

### Release note

Support TopN with ties queries.

1. Normal TopN `SELECT ... ORDER BY ... FETCH FIRST n ROWS WITH TIES` 

2. `RANK` (similar to https://github.com/risingwavelabs/risingwave/pull/5116)

```sql
SELECT .. from
   (SELECT .., RANK() OVER(PARTITION BY .. ORDER BY ..) rank from ..)
WHERE rank < ..;
```

> **Note**: `OFFSET FETCH FIRST n ROWS WITH TIES` and `rank > ..` is not supported yet.

## Refer to a related PR or issue link (optional)
part of https://github.com/risingwavelabs/risingwave/issues/4847

### CODE REVIEW

**Comment by skyzh** [comment]
The `with_ties` bool looks a little bit doubtful to me. If there're ties in topN, we should output a consistent result for all cases. Maybe we should make it by default true (and remove logic w/o tie?)

**Comment by TennyZhuang** [comment]
> The `with_ties` bool looks a little bit doubtful to me. If there're ties in topN, we should output a consistent result for all cases. Maybe we should make it by default true (and remove logic w/o tie?)

It’s used by user:
1. WITH TIES clause
2. rank() in SQL window function

**Comment by skyzh** [comment]
Okay, I misunderstood the feature. I thought it is something like "specify which row to output when there's tie", but it turned out to be output all rows in tie.

File: src/frontend/planner_test/tests/testdata/limit.yaml [hunk 1]
```diff
@@ -45,7 +45,27 @@
...
-  binder_error: |-
-    Feature is not yet implemented: WITH TIES is not supported
+    select * from t order by v fetch next 2 rows with ties;
+  logical_plan: |
+    LogicalTopN { order: "[t.v ASC]", limit: 2, offset: 0, with_ties: true }
+    └─LogicalProject { exprs: [t.v] }
+      └─LogicalScan { table: t, columns: [t.v, t._row_id] }
+  batch_error: |-
+    Feature is not yet implemented: TopN with ties in batch mode
+    Tracking issue: https://github.com/risingwavelabs/risingwave/issues/4848
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/limit.yaml:55]: Linked to a unrelated issue

File: src/frontend/src/optimizer/plan_node/stream_topn.rs [hunk 2]
```diff
@@ -96,7 +101,8 @@ impl StreamNode for StreamTopN {
             ),
             order_by_len: self.topn_order().len() as u32,
         };
-        if self.input().append_only() {
+        // TODO: support with ties for append only TopN
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/stream_topn.rs:104]: Please also create an issue for it. Thanks!

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs [hunk 3]
```diff
@@ -67,7 +67,11 @@ impl Rule for OverAggToTopNRule {
             partition_by,
             order_by,
         } = &over_agg.window_function;
-        assert_eq!(function_type, &WindowFunctionType::RowNumber);
+        let with_ties = match function_type {
+            WindowFunctionType::RowNumber => false,
+            WindowFunctionType::Rank => true,
+            WindowFunctionType::DenseRank => unreachable!("Not implemented. Banned in planenr."),
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/rule/over_agg_to_topn.rs:73]:
```suggestion
            WindowFunctionType::DenseRank => unreachable!("Not implemented. Banned in planner."),
```

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs [hunk 4]
```diff
@@ -118,10 +122,15 @@ impl Rule for OverAggToTopNRule {
             }
         };
 
+        if offset > 0 && with_ties {
+            return None;
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/rule/over_agg_to_topn.rs:126]: Also unreachable here?

File: src/stream/src/executor/group_top_n.rs [hunk 5]
```diff
@@ -196,33 +196,31 @@ where
...
 
             // apply the chunk to state table
             match op {
                 Op::Insert | Op::UpdateInsert => {
                     self.managed_state
                         .insert(ordered_pk_row.clone(), row.clone());
-                    self.caches.get_mut(&pk_prefix.0).unwrap().insert(
-                        ordered_pk_row,
+                    cache.insert(
+                        ordered_pk_row.skip(pk_prefix.size()),
```
Comment (BugenZhao) [inline src/stream/src/executor/group_top_n.rs:223]: Can we directly do this when constructing the `ordered_pk_row` to avoid another allocation? 🥵

File: src/stream/src/executor/group_top_n.rs [hunk 6]
```diff
@@ -423,15 +389,16 @@ mod tests {
         // consume the init barrier
         top_n_executor.next().await.unwrap().unwrap();
         let res = top_n_executor.next().await.unwrap().unwrap();
-        compare_stream_chunk(
+        assert_eq!(
             res.as_chunk().unwrap(),
             &StreamChunk::from_pretty(
                 "  I I I
+                + 10 9 1
```
Comment (BugenZhao) [inline src/stream/src/executor/group_top_n.rs:396]: 👀 Why does the result change?

File: src/stream/src/executor/managed_state/top_n/top_n_state.rs [hunk 7]
```diff
@@ -61,36 +66,38 @@ impl<S: StateStore> ManagedTopNState<S> {
         self.state_table.delete(value);
     }
 
-    fn get_topn_row(&self, row: Row) -> TopNStateRow {
+    fn get_topn_row(&self, row: Row, group_key_len: usize) -> TopNStateRow {
         let mut datums = Vec::with_capacity(self.state_table.pk_indices().len());
-        for pk_index in self.state_table.pk_indices() {
+        for pk_index in self.state_table.pk_indices().iter().skip(group_key_len) {
             datums.push(row[*pk_index].clone());
         }
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/top_n/top_n_state.rs:73]: Should the capacity also be updated, or just `collect` it? BTW, it's really non-sense to first collect an owned row, then transform it to an ordered row. 😅

File: src/stream/src/executor/managed_state/top_n/top_n_state.rs [hunk 8]
```diff
@@ -61,36 +66,38 @@ impl<S: StateStore> ManagedTopNState<S> {
...
+        let state_table_iter = iter_state_table(&self.state_table, group_key).await?;
         pin_mut!(state_table_iter);
 
         // here we don't expect users to have large OFFSET.
-        let (mut rows, mut stream) = if let Some(limits) = num_limit {
+        let (mut rows, mut stream) = if let Some(limit) = limit {
             (
-                Vec::with_capacity(limits.min(1024)),
-                state_table_iter.skip(offset).take(limits),
+                Vec::with_capacity(limit.min(1024)),
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/top_n/top_n_state.rs:99]: Why 1024 here? 🤔

**Review by BugenZhao — APPROVED** [review]
LGTM. Most comments are about the original implementation of Top-N.

File: src/frontend/src/optimizer/rule/over_agg_to_topn.rs (hunk 4 above)
Comment (xxchan) [reply src/frontend/src/optimizer/rule/over_agg_to_topn.rs:126]: It's unfortunately reachable when `RANK() > m`..

File: src/stream/src/executor/managed_state/top_n/top_n_state.rs (hunk 8 above)
Comment (xxchan) [reply src/stream/src/executor/managed_state/top_n/top_n_state.rs:99]: just random stuff as test utility?

File: src/stream/src/executor/group_top_n.rs (hunk 5 above)
Comment (xxchan) [reply src/stream/src/executor/group_top_n.rs:223]: ~~The line above needs it.~~ It actually doesn't need it :think:

File: src/stream/src/executor/managed_state/top_n/top_n_state.rs (hunk 7 above)
Comment (xxchan) [reply src/stream/src/executor/managed_state/top_n/top_n_state.rs:73]: The Row is owned by the TopNStateRow :think:
